`_lib._util.MapWrapper` uses multiprocessing with `fork`, deprecated in Python 3.12

Test failure from the new CI job with Python 3.12 I was just working on:

```
 ___________ TestDifferentialEvolutionSolver.test_immediate_updating ____________
[gw1] linux -- Python 3.12.1 /opt/hostedtoolcache/Python/3.12.1/x64/bin/python
scipy/optimize/tests/test__differential_evolution.py:703: in test_immediate_updating
    with warns(UserWarning):
E   DeprecationWarning: This process (pid=12452) is multi-threaded, use of fork() may lead to deadlocks in the child.
        bounds     = [(0.0, 2.0), (0.0, 2.0)]
        self       = <scipy.optimize.tests.test__differential_evolution.TestDifferentialEvolutionSolver object at 0x7f60117f4e30>
        solver     = <scipy.optimize._differentialevolution.DifferentialEvolutionSolver object at 0x7f5ffc3b0e30>

...

 FAILED scipy/optimize/tests/test__differential_evolution.py::TestDifferentialEvolutionSolver::test_immediate_updating - DeprecationWarning: This process (pid=12452) is multi-threaded, use of fork() may lead to deadlocks in the child.
```

This was inside a `pytest.warns` that was filtering another warning, so this is probably being surfaced now because of the combination of:
- the deprecation which is new in Python 3.12
- use of `pytest` 8.0, which bubbles up more warnings
- running the tests in parallel

Note that the warning isn't deterministic, because Python cannot reliably determine whether a program is using threading. I cannot reproduce it locally, but the cause is clear. See the notes on using the `fork` method with `multiprocessing` in https://docs.python.org/3/whatsnew/3.12.html.

For now I'll just avoid the warning, but I think the thing to do here is to switch the start method to `'spawn'` or `'forkserver'` (which one is TBD - the latter is faster, but not always available).

Cc @andyfaff

--- a/scipy/optimize/tests/test__differential_evolution.py
+++ b/scipy/optimize/tests/test__differential_evolution.py
@@ -709,7 +709,7 @@ class TestDifferentialEvolutionSolver:
             with warns(UserWarning):
                 with DifferentialEvolutionSolver(rosen, bounds, workers=p.map) as s:
                     pass
-            assert_(s._updating == 'deferred')
+            assert s._updating == 'deferred'
 
     def test_parallel(self):
         # smoke test for parallelization with deferred updating

--- a/scipy/optimize/tests/test__differential_evolution.py
+++ b/scipy/optimize/tests/test__differential_evolution.py
@@ -698,12 +698,18 @@ class TestDifferentialEvolutionSolver:
         solver = DifferentialEvolutionSolver(rosen, bounds)
         assert_(solver._updating == 'immediate')
 
-        # should raise a UserWarning because the updating='immediate'
-        # is being overridden by the workers keyword
-        with warns(UserWarning):
-            with DifferentialEvolutionSolver(rosen, bounds, workers=2) as solver:
-                pass
-        assert_(solver._updating == 'deferred')
+        # Safely forking from a multithreaded process is
+        # problematic, and deprecated in Python 3.12, so
+        # we use a slower but portable alternative
+        # see gh-19848
+        ctx = multiprocessing.get_context("spawn")
+        with ctx.Pool(2) as p:
+            # should raise a UserWarning because the updating='immediate'
+            # is being overridden by the workers keyword
+            with warns(UserWarning):
+                with DifferentialEvolutionSolver(rosen, bounds, workers=p.map) as s:
+                    pass
+            assert_(s._updating == 'deferred')
 
     def test_parallel(self):
         # smoke test for parallelization with deferred updating

